<a href="https://colab.research.google.com/github/rahulde0/python-basics/blob/main/Numpy_Operations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Numpy

## Agenda
In this lesson, we will cover the following concepts:
* FUndamentals of Numpy
 * Array Object-> Creating Numpy Arrays

In Python nd-Array can go up to how many dimentions??
Ans: 32 Dimensions. Numpy is written in C language, and the no. of dimensions hardcoded at 32.

In [ ]:
import numpy as np


In [ ]:
arr =  np.array(9)
arr.ndim

0

In [ ]:
arr3d= np.array([[[9,0,1,2,3], [9,2,5,2,6]], [[9,0,1,2,3], [9,2,5,2,6]]])
arr3d.ndim

3

In [ ]:
arr4d = np.array([[[[9]]]])
arr4d.ndim

4

In [ ]:
arr3d.shape

(2, 2, 5)

In [ ]:
# This library is used for working with Arrays
# Consists of multidimensional array of objects and a collecition of funcations for manipulating them
# Conducts mathematical and logical operations.

Adv. of Numpy

* Provides an array object that is faster then traditional Python Lists.
* Numpy are stored in one continouous place in memory unlike lists.

In [ ]:
# Install nunmpy --> pip install numpy

In [ ]:
import numpy as np
arr=np.array([1,2,3,4,5])
print(arr)
print(type(arr))


[1 2 3 4 5]
<class 'numpy.ndarray'>


In [ ]:
list1 = [5,6,7,8,9]
arr1 = np.array(list1)

print(arr1)
print(type(arr1))

[5 6 7 8 9]
<class 'numpy.ndarray'>


In [ ]:
list1 = [5,6,7,8,9]
print(type(list1))
print(list1)

arr1 = np.array(list1)

print(arr1)
print(type(arr1))

<class 'list'>
[5, 6, 7, 8, 9]
[5 6 7 8 9]
<class 'numpy.ndarray'>


In [ ]:
# For accessing numpy array elements, it is same as that we followed in Lists.

Create Numpy Arrays of various dimensions

In [ ]:
#Create 0D, 1D, 2D, 3D
#Create a 0D array:
import numpy as np
print('---------0D array--------------')
arr = np.array(42) # Single numerical value when converted to a numpy array, gives 0D array.
print(arr)

#Create 1D array
print('---------1D array--------------')
arr1 = np.array([1, 2, 3, 4, 5])
print(arr1)

# Create 2D array
print('---------2D array--------------')
arr2 = np.array([[1, 2, 3], [4, 5, 6], [4, 5, 6], [4, 5, 6], [4, 5, 6], [4, 5, 6], [4, 5, 6]]) # Combination of multiple 1D array gives 2D array
print(arr2)

#Create a 3D array
print('---------3D array--------------')
arr3 = np.array([[[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]],[[1, 2, 3], [4, 5, 6]],[[1, 2, 3], [4, 5, 6]],[[1, 2, 3], [4, 5, 6]],[[1, 2, 3], [4, 5, 6]]])
print(arr3)

---------0D array--------------
42
---------1D array--------------
[1 2 3 4 5]
---------2D array--------------
[[1 2 3]
 [4 5 6]
 [4 5 6]
 [4 5 6]
 [4 5 6]
 [4 5 6]
 [4 5 6]]
---------3D array--------------
[[[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]

 [[1 2 3]
  [4 5 6]]]


In [ ]:
list1 = [[1,2,3,"4",6], [1,2,3,4,5]]
arr1 = np.array(list1)
print(arr1)
print(type(arr1))

[['1' '2' '3' '4' '6']
 ['1' '2' '3' '4' '5']]
<class 'numpy.ndarray'>


## Observations:
* 1D arrays are unidimensional and have 0D elements as its elements.
* 2D arrays have 1D arrays as its elements
* 3D arrays have 2D arrays as its elements

In [ ]:
arr = np.array([[[1,2,3,4,5],[1,2,3,4,5],[1,2,3,4,5],[1,2,3,4,5]], [[1,2,3,4,5],[1,2,3,4,5],[1,2,3,4,5],[1,2,3,4,5]]])
print(arr.ndim)

3


In [ ]:
np.array([[[1,2,3,4,4]]]).ndim

3

String Functions in Numpy

In [ ]:
# Perform element wise string concatenation for two araays if string
x= np.array(['Hello', 'Hello'])
y= np.array(['WelcomeHello', 'Learners'])
result = np.char.add(x,y)
result

array(['HelloWelcomeHello', 'HelloLearners'], dtype='<U17')

In [ ]:
result_str = x+y
result_str

array(['HelloWelcomeHello', 'HelloLearners'], dtype='<U17')

In [ ]:
result_str1  = np.add(x,y)
result_str1

array(['HelloWelcomeHello', 'HelloLearners'], dtype='<U17')

In [ ]:
replaced_result = np.char.replace(result, 'Hello', 'Hi' )
print(replaced_result)

['HiWelcomeHi' 'HiLearners']


In [ ]:
replaced_result = np.char.replace(result, 'Hello', 'Hi' )
print(replaced_result)

['HiWelcomeHi' 'HiLearners']


In [ ]:
# append function is not available in npmpy array
#np.char.append(replaced_result, "oooo")

In [ ]:
#upper case and lower case using numpy.char
string1 = 'Hello How Are You'
print(string1)

upper_case = np.char.upper(string1)
print(upper_case)

lower_case = np.char.lower(string1)
print(lower_case)

Hello How Are You
HELLO HOW ARE YOU
hello how are you


Indexing and Slicing in Numpy Array


In [ ]:
#Create a 1d, 2d and 3d array
arr1D = np.array([1,2,3,4,5,6])
arr2D = np.array([[1,2,3,4,5,6],[6,7,5,4,5,0]])
arr3D = np.array([[[1,2,3,4,5,6],[6,7,5,4,5,0]], [[1,2,4,4,5,6],[6,7,5,4,5,9]]])
print('----------1D array------------')
print(arr1D)
print('----------2D array------------')
print(arr2D)
print('----------3D array------------')
print(arr3D)

----------1D array------------
[1 2 3 4 5 6]
----------2D array------------
[[1 2 3 4 5 6]
 [6 7 5 4 5 0]]
----------3D array------------
[[[1 2 3 4 5 6]
  [6 7 5 4 5 0]]

 [[1 2 4 4 5 6]
  [6 7 5 4 5 9]]]


In [ ]:
# Access the 2nd Element in each array
arr1D[1]

np.int64(2)

In [ ]:
# 2nd element of the 2D array --- > 5th Element of the 1D array
(arr2D[1][4])

np.int64(5)

In [ ]:
arr3D[0][1][4]

np.int64(5)

In [ ]:
#Print a 1 D array in reverse
arr1D[::-1]

array([6, 5, 4, 3, 2, 1])

In [ ]:
#Print a 2D array in reverse using indexing and slicing:: assignment-- Leaving this incomplete.
print('Normal seq: ', arr2D)
##print('Rev seq: ', arr2D[::][::-1])

Normal seq:  [[1 2 3 4 5 6]
 [6 7 5 4 5 0]]


In [ ]:
arr3D = np.array([[[1,2,3,4,5,6],[6,7,5,4,5,0]], [[1,2,4,4,5,6],[6,7,5,4,5,9]],  [[9,2,4,6,5,6],[1,7,5,5,5,0]]])
print(arr3D[2][0][-1])

6


In [ ]:
arr3D
# Print this array as 1D array:


array([[[1, 2, 3, 4, 5, 6],
        [6, 7, 5, 4, 5, 0]],

       [[1, 2, 4, 4, 5, 6],
        [6, 7, 5, 4, 5, 9]],

       [[9, 2, 4, 6, 5, 6],
        [1, 7, 5, 5, 5, 0]]])

In [ ]:
arr3D.reshape(36,)

array([1, 2, 3, 4, 5, 6, 6, 7, 5, 4, 5, 0, 1, 2, 4, 4, 5, 6, 6, 7, 5, 4,
       5, 9, 9, 2, 4, 6, 5, 6, 1, 7, 5, 5, 5, 0])

In [ ]:
arr3D.flatten() # Converts a n-d array to 1D array.

array([1, 2, 3, 4, 5, 6, 6, 7, 5, 4, 5, 0, 1, 2, 4, 4, 5, 6, 6, 7, 5, 4,
       5, 9, 9, 2, 4, 6, 5, 6, 1, 7, 5, 5, 5, 0])

In [ ]:
arr3D.shape

(3, 2, 6)

In [ ]:
arr1D = np.reshape(arr3D, -1) # -1 == 36 as part of indexing
arr1D

array([1, 2, 3, 4, 5, 6, 6, 7, 5, 4, 5, 0, 1, 2, 4, 4, 5, 6, 6, 7, 5, 4,
       5, 9, 9, 2, 4, 6, 5, 6, 1, 7, 5, 5, 5, 0])

Assignment: Problem Statement:

Create a 1D NumPy array with at least 10 elements

Create a 2D NumPy array with a minimum of 3 rows and 4 columns

Create a 3D NumPy array with at least 2 matrices, each containing 2 rows and 3 columns

Accessing Elements in Arrays:
Access and print various elements from 1D, 2D, and, 3D using positive indexing

Perform and print some basic arithmetic operations (like addition, subtraction) using elements accessed from 1D, 2D, and, 3D arrays

Access and print elements using negative indices in all three arrays

In [ ]:
list1= [1,2,3,4,5,6,7]
print(id(list1)) # -- > x
list1=[3,4,5,6,7,8] #-->y
print(id(list1))

137594730203840
137594730275072


In [ ]:
x= np.array([[1,2,3,4,5],[3,4,5,56,6]])
y = np.array([1,2,3,4,5,])
print(id(x))
print(id(y))
z = x
print(id(z))


137594730362896
137594731825072
137594730362896


In [ ]:
list2 = [1,2,True, 1.0, 'string']
print(list2)
print(id(list2))

[1, 2, True, 1.0, 'string']
137594729996352


In [ ]:
print(id(list2[0]))
print(id(list2[1]))

11642088
11642120


In [ ]:
print(id(x))

137594730362896


In [ ]:
print(id(x[0]))
print(id(x[1]))

137594731824688
137594731824688


Arithmetic operations Using Numpy

In [ ]:
#Addition
# Any mathematical operations, numpy will try to do element wise operations
x =np.array([1,2,3,4,5])
y=np.array([1])
x+y

array([2, 3, 4, 5, 6])

In [ ]:
#x =np.array([1,2,3,4,5])
#y=np.array([1,2])
#x+y

In [ ]:
x =np.array([1,2,3,4,5])
y=np.array([1,2,3,4,5])
z = x+y # Element wise addition
type(z)
z

array([ 2,  4,  6,  8, 10])

In [ ]:
np.add(x,y)

array([ 2,  4,  6,  8, 10])

In [ ]:
a= np.array([1,2,3,4,5])
b= np.array([1]) # Broadcasting in Numpy. [1,1,1,1,1]
np.add(a,b)

array([2, 3, 4, 5, 6])

In [ ]:
# Substraction, Division, Multiplication, Floor Division, Modulo Division, Power --> Assignment.

Statistical Operations in Numpy
* Mean, Median, Mode, Standard Deviation, Variance in an Array.

In [ ]:
array = np.array([1,2,3,4,5,6,7,8,9,10,11])
np.median(array)

np.float64(6.0)

In [ ]:
np.mean(array)

np.float64(6.0)

In [ ]:
array1 = np.array([1,2,3,4,5,6,7,8,9,10,100])
np.median(array1)

np.float64(6.0)

In [ ]:
np.mean(array1)

np.float64(14.090909090909092)

In [ ]:
np.std(array1)

np.float64(27.304526915561905)

In [ ]:
np.var(array1)

np.float64(745.5371900826445)

In [ ]:
x= range(1,300)
print(x)
list1=list(x)
print(list1)
np.percentile(list1, 100)

range(1, 300)
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219

np.float64(299.0)

In [ ]:
np.percentile(list1, 50)

np.float64(150.0)

In [ ]:
np.median(list1)

np.float64(150.0)

In [ ]:
np.percentile(list1, 25)

np.float64(75.5)

In [ ]:
np.percentile(list1, 75)

np.float64(224.5)

Problem Statement:
Consider the following arrays:
Arrays for Arithmetic Operations:

Array A: [15, 25, 35, 45, 55]

Array B: [5, 10, 15, 20, 25]

Array for Statistical Operations:
Array C: [12, 22, 32, 42, 52, 62, 72, 82, 92, 102]

Arrays for String Functions:
String Array X: ['Hello', 'Data', 'Science', 'Class', 'Learners']
String Array Y: ['World', 'Analysis', 'is', 'Fun', '2023']

Steps to Perform:
Perform element-wise addition, subtraction, multiplication, division, and power operations on Arrays A and B

Calculate the median, mean, standard deviation, and variance for Array C

Concatenate corresponding elements of String Arrays X and Y

Convert all elements in String Array X to uppercase

Replace a specific substring in String Array Y with another substring

In [ ]:
np.random.randint(1,100,10)

array([33, 55, 57, 98, 18, 10, 42, 46, 17, 44])

In [ ]:
np.random.randint(1,100,22)

array([21, 49, 34, 77, 48, 64, 76, 73, 84,  9, 82, 76, 48, 59, 13, 38, 35,
       80, 16, 52, 42, 13])

In [ ]:
np.random.randint(1,3, (5,5))

array([[2, 1, 1, 2, 1],
       [2, 1, 1, 2, 1],
       [1, 1, 1, 2, 1],
       [2, 1, 1, 2, 2],
       [1, 1, 1, 1, 2]])

In [ ]:
#Examples of built-in ufuncs:
#* Vectorized function in numpy, which operates elelemnt-eise on arrays
#Arithmetic: np.add, np.subtract, np.multiply, np.np.divide
#Trigonometric-> np.sin, np.cos, np.tan
#Exponential --> np.exp, np.log,
#np.greater, np.equal


In [ ]:
arr1_1 = np.array([1,2,3,4,5])
np.sin(arr1_1)

array([ 0.84147098,  0.90929743,  0.14112001, -0.7568025 , -0.95892427])

In [ ]:
x=np.array([1,2,3,4,5])
y=np.array([3,4,5,6,7])
np.greater(x,y)

array([False, False, False, False, False])

In [ ]:
x=np.array([1,2,3,4,5])
y=np.array([1,3,3,4,7])
np.equal(x,y)

array([ True, False,  True,  True, False])

In [ ]:
arr2d = np.array(np.random.randint(1,10, (2,5)))
arr2d


array([[6, 4, 6, 8, 9],
       [2, 8, 8, 5, 3]])

In [ ]:
arr2d


array([[6, 4, 6, 8, 9],
       [2, 8, 8, 5, 3]])

In [ ]:
ar =arr2d[::-1][::-1]

In [ ]:
ar[::-1]

array([[6, 4, 6, 8, 9],
       [2, 8, 8, 5, 3]])

In [ ]:
arr3d = np.array(np.random.randint(1,10,(2,3,4)))
arr3d


array([[[9, 7, 6, 4],
        [6, 8, 9, 9],
        [8, 1, 9, 1]],

       [[6, 7, 3, 9],
        [7, 8, 9, 2],
        [9, 8, 3, 5]]])

In [ ]:
arr3d[::-1][::-1][::-1]

array([[[6, 7, 3, 9],
        [7, 8, 9, 2],
        [9, 8, 3, 5]],

       [[9, 7, 6, 4],
        [6, 8, 9, 9],
        [8, 1, 9, 1]]])

In [ ]:
arr3d[::-1]

array([[[6, 7, 3, 9],
        [7, 8, 9, 2],
        [9, 8, 3, 5]],

       [[9, 7, 6, 4],
        [6, 8, 9, 9],
        [8, 1, 9, 1]]])

In [ ]:
print(arr2d)
print('-------------')
print(arr2d[::-1, ::-1])

[[6 4 6 8 9]
 [2 8 8 5 3]]
-------------
[[3 5 8 8 2]
 [9 8 6 4 6]]


Shape of an nd array


In [ ]:
arr0D= np.array(1) # 0 dimentional array has void shape.
print(arr0D)
print(arr0D.shape)

1
()


In [ ]:
arr1D= np.array([1,2,3,4,5,6,7,8,9,10]) # By default the 1D array is 10,1 shape not 1,10.
print(arr1D)
print(arr1D.shape)
arr1D.reshape(1,10)
print(arr1D.reshape(1,10).shape) # Becomes a 2d array.


[ 1  2  3  4  5  6  7  8  9 10]
(10,)
(1, 10)


10,1 --> can be reshaped it to 1,10 >> 2,5 --> 5,2
24,1 --> can be reshaped to --> 2,12 --> 12,2 --> 2,3,4-->
It can be reshaped to any format, just ensure the total no of elements in the array is same.


In [ ]:
arr = np.array([range(1,25)])
print("1 D array:", arr)
print("---------------")
print("2 D array:", arr.reshape(2,12))
print("---------------")
print("3 D array:", arr.reshape(2,3,4))

1 D array: [[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24]]
---------------
2 D array: [[ 1  2  3  4  5  6  7  8  9 10 11 12]
 [13 14 15 16 17 18 19 20 21 22 23 24]]
---------------
3 D array: [[[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]

 [[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]]


In [ ]:
print('3D array:', arr.reshape(4,3,2))

3D array: [[[ 1  2]
  [ 3  4]
  [ 5  6]]

 [[ 7  8]
  [ 9 10]
  [11 12]]

 [[13 14]
  [15 16]
  [17 18]]

 [[19 20]
  [21 22]
  [23 24]]]


In [ ]:
arr3d = np.array([ [[1,2,3,4,5,6,7],[8,9,10,11,12,13,14]], [[11,12,13,14,15,16,17],[18,19,110,111,112,113,114]]])
arr3d

array([[[  1,   2,   3,   4,   5,   6,   7],
        [  8,   9,  10,  11,  12,  13,  14]],

       [[ 11,  12,  13,  14,  15,  16,  17],
        [ 18,  19, 110, 111, 112, 113, 114]]])

In [ ]:
arr3d.shape # 28 elements.
arr3d.reshape(1,28)


array([[  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
         14,  11,  12,  13,  14,  15,  16,  17,  18,  19, 110, 111, 112,
        113, 114]])

In [ ]:
arr3d.flatten()  # Instructing to give me a 1D array.

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  11,  12,  13,  14,  15,  16,  17,  18,  19, 110, 111, 112,
       113, 114])

In [ ]:
arr1D = np.array(range(1,24))
arr1D.reshape(1,1,23)
#print(arr1D.reshape(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,5,5))

array([[[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
         17, 18, 19, 20, 21, 22, 23]]])

In [ ]:
#Checking the data type
arr= np.array([1,2,3,4,5,6,7])
arr.dtype

dtype('int64')

In [ ]:
arr= np.array(['1',2,3,4,5,6,7])
arr.dtype

dtype('<U21')

In [ ]:
arr= np.array(['1',2,3,4,5,6,7,True])
arr.dtype

dtype('<U21')

In [ ]:
arr

array(['1', '2', '3', '4', '5', '6', '7', 'True'], dtype='<U21')

In [ ]:
arr= np.array([2,3,4,5,6,7,True])
arr.dtype

dtype('int64')

In [ ]:
arr

array([2, 3, 4, 5, 6, 7, 1])

In [ ]:
# Array Initialization : Generate arrays with varying step size.
list(range(1,10,2))

[1, 3, 5, 7, 9]

In [ ]:
np.arange(0,10,2)

array([0, 2, 4, 6, 8])

In [ ]:
np.zeros(27).reshape(3,3,3)

array([[[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]]])

In [ ]:
np.zeros(27).reshape(3,3,3)

array([[[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]]])

In [ ]:
arr_1s = np.ones(100).reshape(10,10)
arr_1s.dtype

dtype('float64')

In [ ]:
np.linspace(0,5,11) # Generate 10 numbers, equally spaced betweeen 0 to 5 # starting_num, ending_num, no_of_elements , ending_num is considered.


array([0. , 0.5, 1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. ])

In [ ]:
np.arange(10,0,-1) # starting_num, ending_num, step_size. , ending_num is not printed.

array([10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

In [ ]:
np.eye(3) # Creates an identity matrix.

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [ ]:
#######Assignment####
#1. Find the difference between np.random.rand() and np.random.randn()--> and also its application.

In [ ]:
np.random.rand(10) # Uniform Distribution between 0 t0 1,

array([0.20066823, 0.54604736, 0.34330969, 0.51891662, 0.94519304,
       0.42277377, 0.82683124, 0.05124584, 0.31991402, 0.41459829])

In [ ]:
np.random.randn(1000) # Normal Distribution between -1 to +1 with mean as 0 and standard deviation as 1.

array([-4.24592818e-01,  1.29293492e+00,  1.58126539e+00, -1.17196757e+00,
        1.58533309e-01,  1.49784139e+00,  3.56801592e+00, -1.33461632e+00,
       -6.13207820e-01, -1.69365312e+00, -1.14134379e+00,  9.72510153e-03,
       -4.78551093e-01, -1.82369085e+00,  1.47202806e+00, -1.60720380e+00,
        1.27080540e+00, -1.41168549e+00, -2.93645645e-01, -6.97004224e-01,
        1.50137995e+00,  3.74947052e-01, -3.54108655e-01, -3.62469390e-01,
        9.10937144e-01,  9.34580906e-01, -1.00108379e+00, -2.52722353e-01,
        7.75240786e-02, -1.17362598e+00,  4.26879449e-01,  1.01763666e+00,
       -1.81313243e+00,  2.14891714e-01, -6.54251661e-01, -9.10438205e-01,
        9.43958562e-01,  5.94668922e-01, -1.04575355e+00, -9.32192050e-02,
        6.39636674e-01, -2.29248397e+00, -1.18427408e+00, -4.60747984e-02,
        2.51809951e-01, -5.70350137e-01,  1.58259015e+00,  1.00998268e+00,
       -1.28822877e+00, -6.99827566e-01, -1.26213936e+00, -6.00776581e-01,
        2.20992771e+00,  

Conditional Indexing in an Arrays


In [ ]:
arr1D = np.array([1,2,3,4,5,6,7,8,10])
print(arr1D)

[ 1  2  3  4  5  6  7  8 10]


In [ ]:
np.greater(arr1D,5)

array([False, False, False, False, False,  True,  True,  True,  True])

In [ ]:
arr1D>5

array([False, False, False, False, False,  True,  True,  True,  True])

In [ ]:
print(arr1D>5)

[False False False False False  True  True  True  True]


In [ ]:
arr1D[arr1D>5]

array([ 6,  7,  8, 10])

Array BroadCasting

In [ ]:
arr2d = np.arange(1,13).reshape(4,3)
arr2d

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [ ]:
num=5
arr0d = np.array(num)
arr0d

array(5)

In [ ]:
arr2d +arr0d

array([[ 6,  7,  8],
       [ 9, 10, 11],
       [12, 13, 14],
       [15, 16, 17]])

In [ ]:
arr1d = np.array([1,2,3]) # Default shape is (3,) --> means single row 3 cols.
arr1d1 = np.array([1,2,3,4])

In [ ]:
arr2d + arr1d

array([[ 2,  4,  6],
       [ 5,  7,  9],
       [ 8, 10, 12],
       [11, 13, 15]])

In [ ]:
#arr2d + arr1d1
#no. of columns of 1d array shall be equal
#to that of the main array for broadcasting.

ValueError: operands could not be broadcast together with shapes (4,3) (4,) 

In [ ]:
arr = np.array([1,2,3,4,5,6,7,8,9,10])
max(arr) # Similarly we can get min

np.int64(10)

In [ ]:
#argmax()
arr.argmax() # RETURNS THE INDEX OF THE LARGEST NUMBER IN THE ARRAY.

np.int64(9)

In [ ]:
# Try out other functions like , np.sqrt(), np.exp(),

In [ ]:

arr_x = np.array([[1,2,3,4], [5,6,7,8]])
arr_y = np.array([[21,22,23,24], [25,26,27,28]])


In [ ]:
np.add(arr_x, arr_y)

array([[22, 24, 26, 28],
       [30, 32, 34, 36]])

In [ ]:
#Concetanation
arr_z = np.concatenate((arr_x, arr_y), axis=1) # No of columns will increase //column wise concetanation

In [ ]:
arr_z

array([[ 1,  2,  3,  4, 21, 22, 23, 24],
       [ 5,  6,  7,  8, 25, 26, 27, 28]])

In [ ]:
arr_a = np.concatenate((arr_x, arr_y), axis=0) # no of rows will increase. # row wise concetanation
arr_a

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [21, 22, 23, 24],
       [25, 26, 27, 28]])

In [ ]:
######Horizontal Split and Vertical Split######

In [ ]:
arr_z

array([[ 1,  2,  3,  4, 21, 22, 23, 24],
       [ 5,  6,  7,  8, 25, 26, 27, 28]])

In [ ]:
arr_h_split = np.hsplit(arr_z, 2)
arr_h_split


[array([[1, 2, 3, 4],
        [5, 6, 7, 8]]),
 array([[21, 22, 23, 24],
        [25, 26, 27, 28]])]

In [ ]:
arr_h_split[1]

array([[21, 22, 23, 24],
       [25, 26, 27, 28]])

In [ ]:
arr_h_split[1]

array([[21, 22, 23, 24],
       [25, 26, 27, 28]])

In [ ]:
arr_v_split = np.vsplit(arr_z,2)
arr_v_split

[array([[ 1,  2,  3,  4, 21, 22, 23, 24]]),
 array([[ 5,  6,  7,  8, 25, 26, 27, 28]])]

In [ ]:
arr_v_split[0]

array([[ 1,  2,  3,  4, 21, 22, 23, 24]])

**Conclusion:**
Numpy is open source add on module to Python.

By using NumPy you can speed up your workflow and interface with other packages in the Python ecosystem that use NumPy under the hood.

A growing plethora of scientific and mathematical Python-based packages are using NumPy arrays; though these typically support Python-sequence input, they convert such input to NumPy arrays prior to processing, and they often output NumPy arrays.

It provide common mathematical and numerical routines in pre-compiled, fast functions.

It provides basic routines for manipulating large arrays and matrices of numeric data.

**Key Features**
NumPy arrays have a fixed size decided at the time of creation. Changing the size of an ndarray will create a new array and delete the original.

The elements in a NumPy array are all required to be of the same data type, and thus will be the same size in memory.

NumPy arrays facilitate advanced mathematical and other types of operations on large numbers of data.